From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mechoriet <kevinworm92@gmail.com>
Date: Sun, 23 Apr 2023 11:57:09 +0200
Subject: [PATCH] skip event if no listeners


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 729b8084d9de1dad3e9a6ed25ed4c60134243519..9095ce078739daeee32165fa30812b043e66fdfa 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -555,6 +555,9 @@ public final class SimplePluginManager implements PluginManager {
         HandlerList handlers = event.getHandlers();
         RegisteredListener[] listeners = handlers.getRegisteredListeners();
 
+        if (listeners.length == 0) {
+            return;
+        }
         for (RegisteredListener registration : listeners) {
             if (!registration.getPlugin().isEnabled()) {
                 continue;
