From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hpfxd <me@hpfxd.com>
Date: Sat, 30 Oct 2021 14:42:18 -0400
Subject: [PATCH] Configurable time update frequency.


diff --git a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
index 0e781b78cb37bd85f9f753bd351d7c724a561bd7..023820936a0684c97fefb115e7ab10d35413ce6f 100644
--- a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
+++ b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotWorldConfig.java
@@ -6,6 +6,13 @@ import org.spongepowered.configurate.objectmapping.meta.Comment;
 @ConfigSerializable
 @SuppressWarnings({"FieldCanBeLocal", "FieldMayBeFinal"})
 public class PandaSpigotWorldConfig {
+    @Comment("How many ticks in between sending time updates to players?\n" +
+            "\n" +
+            "The vanilla option is 20 (every second), but PandaSpigot sets the default\n" +
+            "to 100 (every 5 seconds). You would probably be fine setting this even\n" +
+            "higher, unless you're constantly changing the time, or the server is lagging.")
+    public int timeUpdateFrequency = 100;
+    
     @Comment("These options control velocity players receive when damaged.")
     public KnockbackConfig knockback;
     
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index ec145f5551193ee9898c7b344b22fd725c2b054d..12eb4fd70c12a6294ac4ed25fcc61087a4b2d7dc 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -905,7 +905,7 @@ public abstract class MinecraftServer extends com.hpfxd.pandaspigot.tickloop.Ree
             long worldTime = world.getTime();
             final PacketPlayOutUpdateTime worldPacket = new PacketPlayOutUpdateTime(worldTime, dayTime, doDaylight);
             for (EntityHuman entityhuman : world.players) {
-                if (!(entityhuman instanceof EntityPlayer) || (this.ticks + entityhuman.getId()) % 20 != 0) {
+                if (!(entityhuman instanceof EntityPlayer) || (this.ticks + entityhuman.getId()) % world.pandaSpigotConfig.timeUpdateFrequency != 0) {
                     continue;
                 }
                 EntityPlayer entityplayer = (EntityPlayer) entityhuman;
