From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Fri, 3 Jun 2022 20:36:24 -0300
Subject: [PATCH] Allow disabling component selector resolving in books


diff --git a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
index fcf7738c75f21ad23693d903b476c65b4a207a84..8d071fdd1ae8aaa97003f0b56d7e206e6dd4b44f 100644
--- a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
+++ b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
@@ -119,4 +119,6 @@ public class PandaSpigotConfig {
         "https://docs.papermc.io/paper/reference/paper-global-configuration#packet-limiter\n" +
         "(note that \"max-packet-rate\" is renamed to \"maxPacketRate\")")
     public PacketLimiterConfig packetLimiter = PacketLimiterConfig.createDefault();
+
+    public boolean resolveSelectorsInBooks = true;
 }
diff --git a/src/main/java/net/minecraft/server/ItemWrittenBook.java b/src/main/java/net/minecraft/server/ItemWrittenBook.java
index d179049724abe8d5e7a914835d17df59e606f78b..2b84704ac6d0ec0a0bd491b35030f23b2d02f2c2 100644
--- a/src/main/java/net/minecraft/server/ItemWrittenBook.java
+++ b/src/main/java/net/minecraft/server/ItemWrittenBook.java
@@ -49,7 +49,7 @@ public class ItemWrittenBook extends Item {
         if (itemstack != null && itemstack.getTag() != null) {
             NBTTagCompound nbttagcompound = itemstack.getTag();
 
-            if (!nbttagcompound.getBoolean("resolved")) {
+            if (com.hpfxd.pandaspigot.config.PandaSpigotConfig.get().resolveSelectorsInBooks && !nbttagcompound.getBoolean("resolved")) { // PandaSpigot
                 nbttagcompound.setBoolean("resolved", true);
                 if (b(nbttagcompound)) {
                     NBTTagList nbttaglist = nbttagcompound.getList("pages", 8);
