From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Noah van der Aa <ndvdaa@gmail.com>
Date: Tue, 5 Oct 2021 20:04:21 +0200
Subject: [PATCH] Add config option for logging player ip addresses


diff --git a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
index fcf7738c75f21ad23693d903b476c65b4a207a84..1d8e4d1050ebafef8784920481621965fd5c212f 100644
--- a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
+++ b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
@@ -119,4 +119,8 @@ public class PandaSpigotConfig {
         "https://docs.papermc.io/paper/reference/paper-global-configuration#packet-limiter\n" +
         "(note that \"max-packet-rate\" is renamed to \"maxPacketRate\")")
     public PacketLimiterConfig packetLimiter = PacketLimiterConfig.createDefault();
+
+    @Comment("Whether player IP addresses should be logged by the server. This does not impact\n" +
+        "the ability of plugins to log the IP addresses of players.")
+    public boolean logPlayerIpAddresses = true;
 }
diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index b32c64263d6d996008d0aa3b25df1d8d54482aa1..9faa98501275cbd03a34dcc83f281d1e3f40a6e0 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -158,7 +158,10 @@ public class LoginListener implements PacketLoginInListener, IUpdatePlayerListBo
     }
 
     public String d() {
-        return this.i != null ? this.i.toString() + " (" + this.networkManager.getSocketAddress().toString() + ")" : String.valueOf(this.networkManager.getSocketAddress());
+        // PandaSpigot start
+        String ip = com.hpfxd.pandaspigot.config.PandaSpigotConfig.get().logPlayerIpAddresses ? this.networkManager.getSocketAddress().toString() : "<ip address withheld>";
+        return this.i != null ? this.i.toString() + " (" + ip + ")" : ip;
+        // PandaSpigot end
     }
 
     public void a(PacketLoginInStart packetlogininstart) {
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 55e6a1755b7db3411865f1d0ad3b4c38cb66d0a8..9882c599533d8ef52369efb6404a967e736548e0 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -119,7 +119,7 @@ public abstract class PlayerList {
         String s1 = "local";
 
         if (networkmanager.getSocketAddress() != null) {
-            s1 = networkmanager.getSocketAddress().toString();
+            s1 = com.hpfxd.pandaspigot.config.PandaSpigotConfig.get().logPlayerIpAddresses ? networkmanager.getSocketAddress().toString() : "<ip address withheld>"; // PandaSpigot
         }
 
         // Spigot start - spawn location event
diff --git a/src/main/java/net/minecraft/server/ServerConnection.java b/src/main/java/net/minecraft/server/ServerConnection.java
index 7cae1b6df32e289782663732fe9903b8cac89e7c..4974b3730109cf915b0d4f20dcff66bcb1c2e94c 100644
--- a/src/main/java/net/minecraft/server/ServerConnection.java
+++ b/src/main/java/net/minecraft/server/ServerConnection.java
@@ -198,7 +198,7 @@ public class ServerConnection {
                                 throw new ReportedException(crashreport);
                             }
 
-                            ServerConnection.e.warn("Failed to handle packet for " + networkmanager.getSocketAddress(), exception);
+                            ServerConnection.e.warn("Failed to handle packet for " + (com.hpfxd.pandaspigot.config.PandaSpigotConfig.get().logPlayerIpAddresses ? networkmanager.getSocketAddress() : "<ip address withheld>"), exception); // PandaSpigot
                             final ChatComponentText chatcomponenttext = new ChatComponentText("Internal server error");
 
                             networkmanager.a(new PacketPlayOutKickDisconnect(chatcomponenttext), new GenericFutureListener() {
