From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mechoriet <kevinworm92@gmail.com>
Date: Fri, 18 Nov 2022 02:06:42 +0100
Subject: [PATCH] Skip updating entity tracker if server is empty


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index b105de173338011651e19acc9ae3cd3e83765f2a..2012e43510acfe1c9b4b379feaff8fef8774d9aa 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -993,7 +993,9 @@ public abstract class MinecraftServer extends com.hpfxd.pandaspigot.tickloop.Ree
                 }
                 try {
                     // PandaSpigot end
+                if (this.getPlayerList().getPlayerCount() > 0) { // PandaSpigot - Skip updating entity tracker if zero players online
                     worldserver.getTracker().updatePlayers();
+                } // PandaSpigot
                     // PandaSpigot start - controlled flush for entity tracker packets
                 } finally {
                     for (NetworkManager networkManager : disabledFlushes) {
