From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mechoriet <kevinworm92@gmail.com>
Date: Fri, 17 Jun 2022 18:35:40 +0200
Subject: [PATCH] Add getTypeIfLoaded to World


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index b75d78da71f24a0b02bf709d026389add0556489..a084ebc67f4fafce916d2ce02d19a09b30e5d5a4 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -766,6 +766,22 @@ public abstract class World implements IBlockAccess {
         return this.worldProvider.p()[this.getLightLevel(blockposition)];
     }
 
+    // PandaSpigot start - Add getTypeIfLoaded
+    public IBlockData getTypeIfLoaded(BlockPosition blockposition) {
+        if (this.captureTreeGeneration) {
+            for (BlockState previous : this.capturedBlockStates) {
+                if (previous.getX() == blockposition.getX() && previous.getY() == blockposition.getY() && previous.getZ() == blockposition.getZ()) {
+                    return CraftMagicNumbers.getBlock(previous.getTypeId()).fromLegacyData(previous.getRawData());
+                }
+            }
+        }
+        Chunk chunk = this.getChunkIfLoaded(blockposition);
+        if (chunk != null) {
+            return this.isValidLocation(blockposition) ? chunk.getBlockData(blockposition) : Blocks.AIR.getBlockData();
+        }
+        return null;
+    }
+    // PandaSpigot end
     // Spigot start
     public IBlockData getType(BlockPosition blockposition)
     {
