From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mechoriet <kevinworm92@gmail.com>
Date: Fri, 18 Nov 2022 12:38:10 +0100
Subject: [PATCH] Avoid loading chunks in multiple places


diff --git a/src/main/java/net/minecraft/server/BlockChest.java b/src/main/java/net/minecraft/server/BlockChest.java
index eed516e223607e84bb475c51cef21e7f0b1073c2..456bd884a742b684f338b28da4bad0a9ef4ab37f 100644
--- a/src/main/java/net/minecraft/server/BlockChest.java
+++ b/src/main/java/net/minecraft/server/BlockChest.java
@@ -50,7 +50,7 @@ public class BlockChest extends BlockContainer {
         while (iterator.hasNext()) {
             EnumDirection enumdirection = (EnumDirection) iterator.next();
             BlockPosition blockposition1 = blockposition.shift(enumdirection);
-            IBlockData iblockdata1 = world.getType(blockposition1);
+            final IBlockData iblockdata1 = world.getTypeIfLoaded(blockposition1); // PandaSpigot - Use getTypeIfLoaded
 
             if (iblockdata1.getBlock() == this) {
                 this.e(world, blockposition1, iblockdata1);
diff --git a/src/main/java/net/minecraft/server/BlockFire.java b/src/main/java/net/minecraft/server/BlockFire.java
index 76cd512964ce86c8488e0985fd43afc0050bb4a2..37077433406d633b44000b7e2c45ea8c7fa63bcc 100644
--- a/src/main/java/net/minecraft/server/BlockFire.java
+++ b/src/main/java/net/minecraft/server/BlockFire.java
@@ -176,6 +176,7 @@ public class BlockFire extends Block {
                                 }
 
                                 BlockPosition blockposition1 = blockposition.a(j, l, k);
+                                if (!world.isLoaded(blockposition1)) continue; // PandaSpigot - Prevent Fire from loading chunks
                                 int j1 = this.m(world, blockposition1);
 
                                 if (j1 > 0) {
@@ -244,10 +245,14 @@ public class BlockFire extends Block {
     }
 
     private void a(World world, BlockPosition blockposition, int i, Random random, int j) {
-        int k = this.c(world.getType(blockposition).getBlock());
+        // PandaSpigot start - Prevent Fire from loading chunks
+        final IBlockData iblockdata = world.getTypeIfLoaded(blockposition);
+        if (iblockdata == null) return;
+        int k = this.c(iblockdata.getBlock());
+        // PandaSpigot end
 
         if (random.nextInt(i) < k) {
-            IBlockData iblockdata = world.getType(blockposition);
+            //IBlockData iblockdata = world.getType(blockposition); // PandaSpigot - removed since assigned if loaded just above
 
             // CraftBukkit start
             org.bukkit.block.Block theBlock = world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
@@ -304,8 +309,11 @@ public class BlockFire extends Block {
 
             for (int k = 0; k < j; ++k) {
                 EnumDirection enumdirection = aenumdirection[k];
-
-                i = Math.max(this.d(world.getType(blockposition.shift(enumdirection)).getBlock()), i);
+                // PandaSpigot start - only assign if chunk is loaded else skip
+                final IBlockData type = world.getTypeIfLoaded(blockposition.shift(enumdirection));
+                if (type == null) continue;
+                i = Math.max(this.d(type.getBlock()), i);
+                //PandaSpigot end
             }
 
             return i;
diff --git a/src/main/java/net/minecraft/server/BlockLeaves.java b/src/main/java/net/minecraft/server/BlockLeaves.java
index 7bf9018bae77e97ce5d2991897edf186ef1ed741..399c2bddd4e0ed27a98ab07a20d981f39683a4c3 100644
--- a/src/main/java/net/minecraft/server/BlockLeaves.java
+++ b/src/main/java/net/minecraft/server/BlockLeaves.java
@@ -69,7 +69,7 @@ public abstract class BlockLeaves extends BlockTransparent {
                     for (k1 = -b0; k1 <= b0; ++k1) {
                         for (l1 = -b0; l1 <= b0; ++l1) {
                             for (i2 = -b0; i2 <= b0; ++i2) {
-                                Block block = world.getType(blockposition_mutableblockposition.c(j + k1, k + l1, l + i2)).getBlock();
+                                Block block = world.getTypeIfLoaded(blockposition_mutableblockposition.c(j + k1, k + l1, l + i2)).getBlock(); // PandaSpigot - Use getTypeIfLoaded
 
                                 if (block != Blocks.LOG && block != Blocks.LOG2) {
                                     if (block.getMaterial() == Material.LEAVES) {
diff --git a/src/main/java/net/minecraft/server/ChunkCache.java b/src/main/java/net/minecraft/server/ChunkCache.java
index 45d385d31296eda75aa30500118c21406df5fe43..e953d6d80ef020f8e4540eae6881709fc8906b8f 100644
--- a/src/main/java/net/minecraft/server/ChunkCache.java
+++ b/src/main/java/net/minecraft/server/ChunkCache.java
@@ -23,7 +23,7 @@ public class ChunkCache implements IBlockAccess {
 
         for (l = this.a; l <= j; ++l) {
             for (i1 = this.b; i1 <= k; ++i1) {
-                this.c[l - this.a][i1 - this.b] = world.getChunkAt(l, i1);
+                this.c[l - this.a][i1 - this.b] = world.getChunkIfLoaded(l, i1); // PandaSpigot - Don't load chunks for pathfinding
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityEnderman.java b/src/main/java/net/minecraft/server/EntityEnderman.java
index f3afbbd3a20ebf7d51a8fac418cff3e6f59ab56f..a69f9632567b9d8f784b9f63003d6f6f79c17e47 100644
--- a/src/main/java/net/minecraft/server/EntityEnderman.java
+++ b/src/main/java/net/minecraft/server/EntityEnderman.java
@@ -352,7 +352,8 @@ public class EntityEnderman extends EntityMonster {
             int j = MathHelper.floor(this.enderman.locY + random.nextDouble() * 3.0D);
             int k = MathHelper.floor(this.enderman.locZ - 2.0D + random.nextDouble() * 4.0D);
             BlockPosition blockposition = new BlockPosition(i, j, k);
-            IBlockData iblockdata = world.getType(blockposition);
+            IBlockData iblockdata = world.getTypeIfLoaded(blockposition); // PandaSpigot
+            if (iblockdata == null) return; // PandaSpigot
             Block block = iblockdata.getBlock();
 
             if (EntityEnderman.c.contains(block)) {
@@ -386,10 +387,11 @@ public class EntityEnderman extends EntityMonster {
             int j = MathHelper.floor(this.a.locY + random.nextDouble() * 2.0D);
             int k = MathHelper.floor(this.a.locZ - 1.0D + random.nextDouble() * 2.0D);
             BlockPosition blockposition = new BlockPosition(i, j, k);
-            Block block = world.getType(blockposition).getBlock();
+            IBlockData iblockdata = world.getTypeIfLoaded(blockposition); // PandaSpigot - avoid loading chunks
+            if (iblockdata == null) return; // PandaSpigot - avoid loading chunks
             Block block1 = world.getType(blockposition.down()).getBlock();
 
-            if (this.a(world, blockposition, this.a.getCarried().getBlock(), block, block1)) {
+            if (this.a(world, blockposition, this.a.getCarried().getBlock(), iblockdata.getBlock(), block1)) { // PandaSpigot - avoid loading chunks
                 // CraftBukkit start - Place event
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.a, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.a.getCarried().getBlock(), this.a.getCarried().getBlock().toLegacyData(this.a.getCarried())).isCancelled()) {
                 world.setTypeAndData(blockposition, this.a.getCarried(), 3);
diff --git a/src/main/java/net/minecraft/server/Village.java b/src/main/java/net/minecraft/server/Village.java
index e67aa0423f6a28b286174662f29ffc32a74992d9..6ee07e7a8484fc79f79da6a0e688b907382cd1a7 100644
--- a/src/main/java/net/minecraft/server/Village.java
+++ b/src/main/java/net/minecraft/server/Village.java
@@ -36,12 +36,35 @@ public class Village {
         this.k = Lists.newArrayList();
         this.a = world;
     }
+    // PandaSpigot start
+    private BlockPosition[] positions = null;
 
+    private void calculateNewCheckPositions() {
+        if(this.d == null || this.d.equals(BlockPosition.ZERO)) {
+            this.positions = null;
+        } else {
+            this.positions = new BlockPosition[] { this.d.a(-this.e, 0, -this.e),
+                    this.d.a(-this.e, 0, this.e),
+                    this.d.a(this.e, 0, -this.e),
+                    this.d.a(this.e, 0, this.e),
+                    this.d};
+        }
+    }
+    public boolean isVillageAreaLoaded() {
+        for(int i = 0; this.positions != null && i < this.positions.length; i++) {
+            if(this.a.isLoaded(this.positions[i])) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // PandaSpigot end
     public void a(World world) {
         this.a = world;
     }
 
     public void a(int i) {
+        if(!this.isVillageAreaLoaded()) { return; } // PandaSpigot - avoid loading chunks
         this.g = i;
         this.m();
         this.l();
@@ -341,6 +364,7 @@ public class Village {
 
             this.e = Math.max(32, (int) Math.sqrt((double) j) + 1);
         }
+        this.calculateNewCheckPositions(); // PandaSpigot
     }
 
     public int a(String s) {
@@ -395,6 +419,7 @@ public class Village {
                 this.j.put(nbttagcompound2.getString("Name"), Integer.valueOf(nbttagcompound2.getInt("S")));
             }
         }
+        this.calculateNewCheckPositions(); // PandaSpigot
 
     }
 
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index a92b1eb303e01bc42157fe384279d7ad42be03f8..9a7aa50a815d15f3e221a9f0883ad128275f2a87 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2708,8 +2708,11 @@ public abstract class World implements IBlockAccess {
 
         for (int i1 = i; i1 <= j; ++i1) {
             for (int j1 = k; j1 <= l; ++j1) {
-                if (this.isChunkLoaded(i1, j1, true)) {
-                    this.getChunkAt(i1, j1).a(entity, axisalignedbb, arraylist, predicate);
+                // PandaSpigot start - Use getChunkIfLoaded
+                Chunk chunk = this.getChunkIfLoaded(i1, j1);
+                if(chunk != null) {
+                    chunk.a(entity, axisalignedbb, arraylist, predicate);
+                // PandaSpigot end
                 }
             }
         }
