From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mechoriet <kevinworm92@gmail.com>
Date: Sun, 4 Dec 2022 12:20:58 +0100
Subject: [PATCH] Don't save empty scoreboard teams to scoreboard.dat


diff --git a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
index ae3cdaca275b1a50c5c37c3dd0021c4bd579d373..d81547a0accf9f1e2af7482effd0e9638ff341e2 100644
--- a/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
+++ b/src/main/java/com/hpfxd/pandaspigot/config/PandaSpigotConfig.java
@@ -125,4 +125,8 @@ public class PandaSpigotConfig {
     @Comment("Whether player IP addresses should be logged by the server. This does not impact\n" +
         "the ability of plugins to log the IP addresses of players.")
     public boolean logPlayerIpAddresses = true;
+
+    @Comment("Disable the saving of empty scoreboard teams sometimes bad plugins could try to keep these resulting in 100's of teams that have 0 use and just wasted space\n" +
+        "and could even impact team lookups if this grows out of bounds")
+    public boolean saveEmptyScoreboardTeams = false;
 }
diff --git a/src/main/java/net/minecraft/server/PersistentScoreboard.java b/src/main/java/net/minecraft/server/PersistentScoreboard.java
index 370c1b31246045e6a864fd077cd9bbf0c65a2ab0..c4eb44c41192e68711fdc8eefdec0441177c0305 100644
--- a/src/main/java/net/minecraft/server/PersistentScoreboard.java
+++ b/src/main/java/net/minecraft/server/PersistentScoreboard.java
@@ -174,6 +174,7 @@ public class PersistentScoreboard extends PersistentBase {
 
         while (iterator.hasNext()) {
             ScoreboardTeam scoreboardteam = (ScoreboardTeam) iterator.next();
+            if (!com.hpfxd.pandaspigot.config.PandaSpigotConfig.get().saveEmptyScoreboardTeams && scoreboardteam.getPlayerNameSet().isEmpty()) continue; // PandaSpigot - setting to not save empty teams
             NBTTagCompound nbttagcompound = new NBTTagCompound();
 
             nbttagcompound.setString("Name", scoreboardteam.getName());
