From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mezz <tehgeek@gmail.com>
Date: Thu, 24 Nov 2022 12:13:45 -0300
Subject: [PATCH] Fix MC-117075: TE Unload Lag Spike


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1e401be460c903450b9143cb5dfc8afc89db0390..999ed6b4a7d684d677faec3e34a01b07b618095e 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1679,7 +1679,11 @@ public abstract class World implements IBlockAccess {
         this.M = true;
         // CraftBukkit start - From below, clean up tile entities before ticking them
         if (!this.c.isEmpty()) {
-            this.tileEntityList.removeAll(this.c);
+            // PandaSpigot start - Use alternate implementation with faster contains
+            java.util.Set<TileEntity> toRemove = java.util.Collections.newSetFromMap(new java.util.IdentityHashMap<>());
+            toRemove.addAll(c);
+            this.tileEntityList.removeAll(toRemove);
+            // PandaSpigot end
             //this.h.removeAll(this.c); // PaperSpigot - Remove unused list
             this.c.clear();
         }
