From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Sat, 22 Jun 2024 15:56:45 -0300
Subject: [PATCH] Improve portal search logic


diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 72f3735b196cbd7d420c6338eb7aba01b40c7804..486796d47cce34c94046548eee11b715c64c3f1d 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -13,7 +13,7 @@ import org.bukkit.util.Vector;
 
 public class PortalTravelAgent {
 
-    private final WorldServer a;
+    protected final WorldServer a; // PandaSpigot - private -> protected
     private final Random b;
     private final LongHashMap<PortalTravelAgent.ChunkCoordinatesPortal> c = new LongHashMap();
     private final List<Long> d = Lists.newArrayList();
@@ -30,9 +30,13 @@ public class PortalTravelAgent {
                 this.b(entity, f);
             }
         } else {
+            // PandaSpigot start - Moved down
+            /*
             int i = MathHelper.floor(entity.locX);
             int j = MathHelper.floor(entity.locY) - 1;
             int k = MathHelper.floor(entity.locZ);
+            */
+            // PandaSpigot end
             // CraftBukkit start - Modularize end portal creation
             BlockPosition created = this.createEndPortal(entity.locX, entity.locY, entity.locZ);
             entity.setPositionRotation((double) created.getX(), (double) created.getY(), (double) created.getZ(), entity.yaw, 0.0F);
@@ -63,7 +67,7 @@ public class PortalTravelAgent {
             }
 
         // CraftBukkit start
-        return new BlockPosition(i, k, k);
+        return new BlockPosition(i, j, k); // PandaSpigot - Fix createEndPortal locY
     }
 
     // use logic based on creation to verify end portal
@@ -94,7 +98,7 @@ public class PortalTravelAgent {
 
     public boolean b(Entity entity, float f) {
         // CraftBukkit start - Modularize portal search process and entity teleportation
-        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, 128);
+        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, a.paperSpigotConfig.portalSearchRadius); // PandaSpigot - Respect the Paper portal search radius configuration
         if (found == null) {
             return false;
         }
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index d68deb81f9f7dfb82591d5876daaac9a3a258dc6..47af49c25e1ce57a93aa4797f833ef8381fab523 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -67,7 +67,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         this.manager = new PlayerChunkMap(this, spigotConfig.viewDistance); // Spigot
         this.worldProvider.a(this);
         this.chunkProvider = this.k();
-        this.Q = ((org.bukkit.craftbukkit.CraftTravelAgent) new org.bukkit.craftbukkit.CraftTravelAgent(this).setSearchRadius(paperSpigotConfig.portalSearchRadius)); // CraftBukkit // Paper - configurable search radius
+        this.Q = new org.bukkit.craftbukkit.CraftTravelAgent(this); // CraftBukkit // Paper - configurable search radius // PandaSpigot - Just define the variable directly
         this.B();
         this.C();
         this.getWorldBorder().a(minecraftserver.aI());
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
index cae65cf86fac24733b9732a6c66bed1e7b45a8a9..a9bda3946cc4b2364f87364aadbdd227e14023b8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
@@ -11,7 +11,7 @@ public class CraftTravelAgent extends PortalTravelAgent implements TravelAgent {
 
     public static TravelAgent DEFAULT = null;
 
-    private int searchRadius = 128;
+    private int searchRadius = a.paperSpigotConfig.portalSearchRadius; // PandaSpigot - Configurable portal search radius
     private int creationRadius = 16;
     private boolean canCreatePortal = true;
 
