From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Sat, 22 Jun 2024 15:56:45 -0300
Subject: [PATCH] Improve portal search logic

Co-authored-by: Poweruser <poweruser.rs@hotmail.com>

diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 72f3735b196cbd7d420c6338eb7aba01b40c7804..fa0ad789e58255d3faa01deb5bc15d517790e69a 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -13,7 +13,7 @@ import org.bukkit.util.Vector;
 
 public class PortalTravelAgent {
 
-    private final WorldServer a;
+    protected final WorldServer a; // PandaSpigot - private -> protected
     private final Random b;
     private final LongHashMap<PortalTravelAgent.ChunkCoordinatesPortal> c = new LongHashMap();
     private final List<Long> d = Lists.newArrayList();
@@ -94,7 +94,13 @@ public class PortalTravelAgent {
 
     public boolean b(Entity entity, float f) {
         // CraftBukkit start - Modularize portal search process and entity teleportation
-        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, 128);
+        // PandaSpigot start - Search proximity first and then do the larger search if the proximity search fails
+        // TODO: If a player sets the portal search radius to some value equal to or less than 10 we can use normal logic
+        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, 10);
+        if (found == null) {
+            found = this.findPortal(entity.locX, entity.locY, entity.locZ, a.paperSpigotConfig.portalSearchRadius); // PandaSpigot - Respect the Paper portal search radius configuration
+        }
+        // PandaSpigot end
         if (found == null) {
             return false;
         }
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index d68deb81f9f7dfb82591d5876daaac9a3a258dc6..47af49c25e1ce57a93aa4797f833ef8381fab523 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -67,7 +67,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         this.manager = new PlayerChunkMap(this, spigotConfig.viewDistance); // Spigot
         this.worldProvider.a(this);
         this.chunkProvider = this.k();
-        this.Q = ((org.bukkit.craftbukkit.CraftTravelAgent) new org.bukkit.craftbukkit.CraftTravelAgent(this).setSearchRadius(paperSpigotConfig.portalSearchRadius)); // CraftBukkit // Paper - configurable search radius
+        this.Q = new org.bukkit.craftbukkit.CraftTravelAgent(this); // CraftBukkit // Paper - configurable search radius // PandaSpigot - Just define the variable directly
         this.B();
         this.C();
         this.getWorldBorder().a(minecraftserver.aI());
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
index cae65cf86fac24733b9732a6c66bed1e7b45a8a9..ba494d07d25ebb3296c0c0d8e2500c655d6c8288 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
@@ -11,7 +11,7 @@ public class CraftTravelAgent extends PortalTravelAgent implements TravelAgent {
 
     public static TravelAgent DEFAULT = null;
 
-    private int searchRadius = 128;
+    private int searchRadius = a.paperSpigotConfig.portalSearchRadius; // PandaSpigot - Configurable portal search radius
     private int creationRadius = 16;
     private boolean canCreatePortal = true;
 
@@ -44,7 +44,13 @@ public class CraftTravelAgent extends PortalTravelAgent implements TravelAgent {
     @Override
     public Location findPortal(Location location) {
         PortalTravelAgent pta = ((CraftWorld) location.getWorld()).getHandle().getTravelAgent();
-        BlockPosition found = pta.findPortal(location.getX(), location.getY(), location.getZ(), this.getSearchRadius());
+        // PandaSpigot start - Search proximity first and then do the larger search if the proximity search fails
+        // TODO: If a player sets the portal search radius to some value equal to or less than 10 we can use normal logic
+        BlockPosition found = pta.findPortal(location.getX(), location.getY(), location.getZ(), 10);
+        if (found == null) {
+            found = pta.findPortal(location.getX(), location.getY(), location.getZ(), this.getSearchRadius());
+        }
+        // PandaSpigot end
         return found != null ? new Location(location.getWorld(), found.getX(), found.getY(), found.getZ(), location.getYaw(), location.getPitch()) : null;
     }
 
