From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Thu, 4 May 2023 14:27:42 -0300
Subject: [PATCH] Fix wrong class on paper hopper patch


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 9c22cf970288d24a431df836409fc8026dee2384..eb67efbf7bb7d175ebdf0b69f2e26e4cfd26164d 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -1034,7 +1034,7 @@ public class Chunk {
 
         // PaperSpigot start
         int[] counts;
-        if (ItemStack.class.isAssignableFrom(oclass)) {
+        if (EntityItem.class.isAssignableFrom(oclass)) { // PandaSpigot - ItemStack -> EntityItem
             counts = itemCounts;
         } else if (IInventory.class.isAssignableFrom(oclass)) {
             counts = inventoryEntityCounts;
