From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mechoriet <kevinworm92@gmail.com>
Date: Fri, 6 Oct 2023 21:18:05 +0200
Subject: [PATCH] Implement EntityPickupItemEvent


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 1f8b1c726ef7343ee7feb1e4ea4c1ef8cc453723..61ad7b545196ebcf5a72f90505097a3cadea6a30 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -317,6 +317,14 @@ public class EntityItem extends Entity {
                 if (event.isCancelled()) {
                     return;
                 }
+                // Call newer event afterwards
+                org.bukkit.event.entity.EntityPickupItemEvent entityEvent = new org.bukkit.event.entity.EntityPickupItemEvent(entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.world.getServer().getPluginManager().callEvent(entityEvent);
+                if (entityEvent.isCancelled()) {
+                    return;
+                }
+                itemstack.count = canHold + remaining;
 
                 // Possibly < 0; fix here so we do not have to modify code below
                 this.pickupDelay = 0;
