From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Wed, 30 Oct 2024 17:56:59 -0300
Subject: [PATCH] Remove unnecessary permgen size check


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 598e9d06e5491eb79cc5f253ca9e1de7c025eec6..1d64ed78786912749156a0a853b9e9787a4fec3a 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -208,6 +208,8 @@ public class Main {
                     // PandaSpigot end
                 }
 
+                // PandaSpigot start - Not necessary to check permgen size anymore
+                /*
                 // Spigot Start
                 int maxPermGen = 0; // In kb
                 for ( String s : java.lang.management.ManagementFactory.getRuntimeMXBean().getInputArguments() )
@@ -224,6 +226,8 @@ public class Main {
                     System.out.println( "Please see http://www.spigotmc.org/wiki/changing-permgen-size/ for more details and more in-depth instructions." );
                 }
                 // Spigot End
+                */
+                // PandaSpigot end
                 System.setProperty("library.jansi.version", "PandaSpigot"); // PandaSpigot - set meaningless jansi version to prevent git builds from crashing on Windows
                 System.out.println("Loading libraries, please wait...");
                 // PandaSpigot start - Modern tick loop
