From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ptbnate <nasxeaty@gmail.com>
Date: Tue, 11 Feb 2025 10:41:32 +0500
Subject: [PATCH] Fix entities going out of sight


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index b4f301a3eb660b9bf080c4f6b4e3bbaa3678a8d6..a4364ccb14038d40450abc7d17ed102424445b9d 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -296,6 +296,21 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     }
 
+    // PandaSpigot start - Fix players going out of sight
+    @Override
+    public boolean hasLineOfSight(Entity entity) {
+        final Vec3D vec = new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ);
+
+        for (int i = 1; i < 3; i++) {
+            double targetY = entity.locY + (entity.getHeadHeight() / 3) * i;
+            if (this.world.rayTrace(vec, new Vec3D(entity.locX, targetY, entity.locZ)) == null) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // PandaSpigot end
+
     public void l() {
         try {
             super.t_();
