From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Thu, 24 Apr 2025 17:02:53 -0300
Subject: [PATCH] Fix wrong handler name used for cast to PacketCompressor


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index f18408e8b16dab15af8c0af26805d3a2f374a37b..4d6a8ec119369a93929be40e9cbaa5d5f0d1f4bc 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -497,7 +497,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
             }
 
             if (this.channel.pipeline().get("compress") instanceof PacketCompressor) {
-                ((PacketCompressor) this.channel.pipeline().get("decompress")).a(i);
+                ((PacketCompressor) this.channel.pipeline().get("compress")).a(i); // PandaSpigot - Fix wrong handler name used for cast to PacketCompressor
             } else {
                 this.channel.pipeline().addBefore("encoder", "compress", new PacketCompressor(i));
             }
