From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MasterDash5 <constant4337@molecularmail.com>
Date: Mon, 22 Sep 2025 19:04:01 -0600
Subject: [PATCH] Fix MC-120567: Bed portal crash


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 99eb33eef6e11f76385afd536a921e5cffa21cb8..864b63f20167491ecef7e51be7d170d8b141d166 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -314,14 +314,11 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
     protected void e() {
         this.O = false;
-        Iterator iterator = this.players.iterator();
+        Iterator iterator = this.players.stream().filter(EntityHuman::isSleeping).iterator(); // PandaSpigot - Fix MC-120567
 
         while (iterator.hasNext()) {
             EntityHuman entityhuman = (EntityHuman) iterator.next();
-
-            if (entityhuman.isSleeping()) {
-                entityhuman.a(false, false, true);
-            }
+            entityhuman.a(false, false, true); // PandaSpigot - Fix MC-120567
         }
 
         this.ag();
