From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mechoriet <kevinworm92@gmail.com>
Date: Sat, 7 Oct 2023 23:26:59 +0200
Subject: [PATCH]  tempfixup! Add packet limiter config


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 5c631068db79615bf51ed357bc9bda2b0ebb0c3b..87ecc1b9a46bfcc146b7b42f142cac46c02faa9c 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -126,7 +126,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
             this.close(reason[0]);
             this.k();
             this.stopReadingPackets = true;
-        }, null);
+        }, (GenericFutureListener<? extends Future<? super Void>>) null);
     }
     // PandaSpigot end - packet limiter
     public NetworkManager(EnumProtocolDirection enumprotocoldirection) {
