From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DieUnterKreatur <info@dieunterkreatur.de>
Date: Tue, 7 Apr 2026 20:28:31 +0200
Subject: [PATCH] Backport-CreateExplosion


diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index 6671d4920f4b808d61fd458ef3f26f21a0e96322..d1d36a9e80ace36ef1cfbb627d2d181d5389ffaa 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -30,6 +30,7 @@ public class Explosion {
     private final List<BlockPosition> blocks = Lists.newArrayList();
     private final Map<EntityHuman, Vec3D> k = Maps.newHashMap();
     public boolean wasCanceled = false; // CraftBukkit - add field
+    public boolean excludeSourceFromDamage = false; // PandaSpigot
 
     public Explosion(World world, Entity entity, double d0, double d1, double d2, float f, boolean flag, boolean flag1) {
         this.world = world;
@@ -41,6 +42,19 @@ public class Explosion {
         this.a = flag;
         this.b = flag1;
     }
+    // PandaSpigot start
+    public Explosion(World world, Entity entity, double d0, double d1, double d2, float f, boolean flag, boolean flag1, boolean excludeSourceFromDamage) {
+        this.world = world;
+        this.source = entity;
+        this.size = (float) Math.max(f, 0.0); // CraftBukkit - clamp bad values
+        this.posX = d0;
+        this.posY = d1;
+        this.posZ = d2;
+        this.a = flag;
+        this.b = flag1;
+        this.excludeSourceFromDamage = excludeSourceFromDamage;
+    }
+    // PandaSpigot stop
 
     public void a() {
         // CraftBukkit start
@@ -108,7 +122,7 @@ public class Explosion {
         int j1 = MathHelper.floor(this.posZ - (double) f3 - 1.0D);
         int k1 = MathHelper.floor(this.posZ + (double) f3 + 1.0D);
         // PaperSpigot start - Fix lag from explosions processing dead entities
-        List list = this.world.a(this.source, new AxisAlignedBB((double) i, (double) l, (double) j1, (double) j, (double) i1, (double) k1), new com.google.common.base.Predicate<Entity>() {
+        List list = this.world.a(excludeSourceFromDamage ? this.source : (Entity) null, new AxisAlignedBB((double) i, (double) l, (double) j1, (double) j, (double) i1, (double) k1), new com.google.common.base.Predicate<Entity>() { // PandaSpigot
             @Override
             public boolean apply(Entity entity) {
                 return IEntitySelector.d.apply(entity) && !entity.dead;
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index fe9b4ea7584587c4a4d198e01fc21838ec59f471..3d5e2db11ddb4e8fb0eb14a5a3c9b262bc15aabc 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2040,17 +2040,22 @@ public abstract class World implements IBlockAccess {
 
         return false;
     }
+    // PandaSpigot start
+    public Explosion createExplosion(Entity entity, double d0, double d1, double d2, float f, boolean flag, boolean flag1, boolean excludeSourceFromDamage) {
+        Explosion explosion = new Explosion(this, entity, d0, d1, d2, f, flag, flag1, excludeSourceFromDamage);
+
+        explosion.a();
+        explosion.a(true);
+        return explosion;
+    }
+    // PandaSpigot stop
 
     public Explosion explode(Entity entity, double d0, double d1, double d2, float f, boolean flag) {
-        return this.createExplosion(entity, d0, d1, d2, f, false, flag);
+        return this.createExplosion(entity, d0, d1, d2, f, false, flag, false); // PandaSpigot
     }
 
     public Explosion createExplosion(Entity entity, double d0, double d1, double d2, float f, boolean flag, boolean flag1) {
-        Explosion explosion = new Explosion(this, entity, d0, d1, d2, f, flag, flag1);
-
-        explosion.a();
-        explosion.a(true);
-        return explosion;
+        return this.createExplosion(entity, d0, d1, d2, f, flag, flag1, false); // PandaSpigot 
     }
 
     public float a(Vec3D vec3d, AxisAlignedBB axisalignedbb) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b1e389f16a2d5a85a15bb934117f4ba207c856b9..fe0709ff03e497315ee71aa8c5f49c1b438896b1 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -133,9 +133,11 @@ public class CraftWorld implements World {
             }
         });
     }
+
     public void getChunkAtAsync(Block block, ChunkLoadCallback callback) {
         getChunkAtAsync(block.getX() >> 4, block.getZ() >> 4, callback);
     }
+
     public void getChunkAtAsync(Location location, ChunkLoadCallback callback) {
         getChunkAtAsync(location.getBlockX() >> 4, location.getBlockZ() >> 4, callback);
     }
@@ -186,7 +188,7 @@ public class CraftWorld implements World {
     }
 
     public boolean unloadChunkRequest(int x, int z, boolean safe) {
-        org.spigotmc.AsyncCatcher.catchOp( "chunk unload"); // Spigot
+        org.spigotmc.AsyncCatcher.catchOp("chunk unload"); // Spigot
         if (safe && isChunkInUse(x, z)) {
             return false;
         }
@@ -197,7 +199,7 @@ public class CraftWorld implements World {
     }
 
     public boolean unloadChunk(int x, int z, boolean save, boolean safe) {
-        org.spigotmc.AsyncCatcher.catchOp( "chunk unload"); // Spigot
+        org.spigotmc.AsyncCatcher.catchOp("chunk unload"); // Spigot
         if (safe && isChunkInUse(x, z)) {
             return false;
         }
@@ -208,11 +210,13 @@ public class CraftWorld implements World {
             return false;
         }
         // PaperSpigot end
-        if (chunk.mustSave) {   // If chunk had previously been queued to save, must do save to avoid loss of that data
+        if (chunk.mustSave) { // If chunk had previously been queued to save, must do save to avoid loss of
+                              // that data
             save = true;
         }
 
-        chunk.removeEntities(); // Always remove entities - even if discarding, need to get them out of world table
+        chunk.removeEntities(); // Always remove entities - even if discarding, need to get them out of world
+                                // table
 
         if (save && !(chunk instanceof EmptyChunk)) {
             world.chunkProviderServer.saveChunk(chunk);
@@ -253,9 +257,11 @@ public class CraftWorld implements World {
         int px = x << 4;
         int pz = z << 4;
 
-        // If there are more than 64 updates to a chunk at once, it will update all 'touched' sections within the chunk
+        // If there are more than 64 updates to a chunk at once, it will update all
+        // 'touched' sections within the chunk
         // And will include biome data if all sections have been 'touched'
-        // This flags 65 blocks distributed across all the sections of the chunk, so that everything is sent, including biomes
+        // This flags 65 blocks distributed across all the sections of the chunk, so
+        // that everything is sent, including biomes
         int height = getMaxHeight() / 16;
         for (int idx = 0; idx < 64; idx++) {
             world.notify(new BlockPosition(px + (idx / height), ((idx % height) * 16), pz));
@@ -270,7 +276,7 @@ public class CraftWorld implements World {
     }
 
     public boolean loadChunk(int x, int z, boolean generate) {
-        org.spigotmc.AsyncCatcher.catchOp( "chunk load"); // Spigot
+        org.spigotmc.AsyncCatcher.catchOp("chunk load"); // Spigot
         chunkLoadCount++;
         if (generate) {
             // Use the default variant of loadChunk when generate == true.
@@ -303,7 +309,8 @@ public class CraftWorld implements World {
                         continue;
                     }
 
-                    net.minecraft.server.Chunk neighbor = world.chunkProviderServer.getChunkIfLoaded(chunk.locX + x, chunk.locZ + z);
+                    net.minecraft.server.Chunk neighbor = world.chunkProviderServer.getChunkIfLoaded(chunk.locX + x,
+                            chunk.locZ + z);
                     if (neighbor != null) {
                         neighbor.setNeighborLoaded(-x, -z);
                         chunk.setNeighborLoaded(x, z);
@@ -336,7 +343,8 @@ public class CraftWorld implements World {
         entity.pickupDelay = 10;
         world.addEntity(entity);
         // TODO this is inconsistent with how Entity.getBukkitEntity() works.
-        // However, this entity is not at the moment backed by a server entity class so it may be left.
+        // However, this entity is not at the moment backed by a server entity class so
+        // it may be left.
         return new CraftItem(world.getServer(), entity);
     }
 
@@ -417,63 +425,69 @@ public class CraftWorld implements World {
     public boolean generateTree(Location loc, TreeType type) {
         net.minecraft.server.WorldGenerator gen;
         switch (type) {
-        case BIG_TREE:
-            gen = new WorldGenBigTree(true);
-            break;
-        case BIRCH:
-            gen = new WorldGenForest(true, false);
-            break;
-        case REDWOOD:
-            gen = new WorldGenTaiga2(true);
-            break;
-        case TALL_REDWOOD:
-            gen = new WorldGenTaiga1();
-            break;
-        case JUNGLE:
-            IBlockData iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
-            IBlockData iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE).set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
-            gen = new WorldGenJungleTree(true, 10, 20, iblockdata1, iblockdata2); // Magic values as in BlockSapling
-            break;
-        case SMALL_JUNGLE:
-            iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
-            iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE).set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
-            gen = new WorldGenTrees(true, 4 + rand.nextInt(7), iblockdata1, iblockdata2, false);
-            break;
-        case COCOA_TREE:
-            iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
-            iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE).set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
-            gen = new WorldGenTrees(true, 4 + rand.nextInt(7), iblockdata1, iblockdata2, true);
-            break;
-        case JUNGLE_BUSH:
-            iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
-            iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.OAK).set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
-            gen = new WorldGenGroundBush(iblockdata1, iblockdata2);
-            break;
-        case RED_MUSHROOM:
-            gen = new WorldGenHugeMushroom(Blocks.RED_MUSHROOM_BLOCK);
-            break;
-        case BROWN_MUSHROOM:
-            gen = new WorldGenHugeMushroom(Blocks.BROWN_MUSHROOM_BLOCK);
-            break;
-        case SWAMP:
-            gen = new WorldGenSwampTree();
-            break;
-        case ACACIA:
-            gen = new WorldGenAcaciaTree(true);
-            break;
-        case DARK_OAK:
-            gen = new WorldGenForestTree(true);
-            break;
-        case MEGA_REDWOOD:
-            gen = new WorldGenMegaTree(false, rand.nextBoolean());
-            break;
-        case TALL_BIRCH:
-            gen = new WorldGenForest(true, true);
-            break;
-        case TREE:
-        default:
-            gen = new WorldGenTrees(true);
-            break;
+            case BIG_TREE:
+                gen = new WorldGenBigTree(true);
+                break;
+            case BIRCH:
+                gen = new WorldGenForest(true, false);
+                break;
+            case REDWOOD:
+                gen = new WorldGenTaiga2(true);
+                break;
+            case TALL_REDWOOD:
+                gen = new WorldGenTaiga1();
+                break;
+            case JUNGLE:
+                IBlockData iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT,
+                        BlockWood.EnumLogVariant.JUNGLE);
+                IBlockData iblockdata2 = Blocks.LEAVES.getBlockData()
+                        .set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE)
+                        .set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
+                gen = new WorldGenJungleTree(true, 10, 20, iblockdata1, iblockdata2); // Magic values as in BlockSapling
+                break;
+            case SMALL_JUNGLE:
+                iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
+                iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE)
+                        .set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
+                gen = new WorldGenTrees(true, 4 + rand.nextInt(7), iblockdata1, iblockdata2, false);
+                break;
+            case COCOA_TREE:
+                iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
+                iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.JUNGLE)
+                        .set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
+                gen = new WorldGenTrees(true, 4 + rand.nextInt(7), iblockdata1, iblockdata2, true);
+                break;
+            case JUNGLE_BUSH:
+                iblockdata1 = Blocks.LOG.getBlockData().set(BlockLog1.VARIANT, BlockWood.EnumLogVariant.JUNGLE);
+                iblockdata2 = Blocks.LEAVES.getBlockData().set(BlockLeaves1.VARIANT, BlockWood.EnumLogVariant.OAK)
+                        .set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
+                gen = new WorldGenGroundBush(iblockdata1, iblockdata2);
+                break;
+            case RED_MUSHROOM:
+                gen = new WorldGenHugeMushroom(Blocks.RED_MUSHROOM_BLOCK);
+                break;
+            case BROWN_MUSHROOM:
+                gen = new WorldGenHugeMushroom(Blocks.BROWN_MUSHROOM_BLOCK);
+                break;
+            case SWAMP:
+                gen = new WorldGenSwampTree();
+                break;
+            case ACACIA:
+                gen = new WorldGenAcaciaTree(true);
+                break;
+            case DARK_OAK:
+                gen = new WorldGenForestTree(true);
+                break;
+            case MEGA_REDWOOD:
+                gen = new WorldGenMegaTree(false, rand.nextBoolean());
+                break;
+            case TALL_BIRCH:
+                gen = new WorldGenForest(true, true);
+                break;
+            case TREE:
+            default:
+                gen = new WorldGenTrees(true);
+                break;
         }
 
         return gen.generate(world, rand, new BlockPosition(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
@@ -494,7 +508,7 @@ public class CraftWorld implements World {
                 net.minecraft.server.Block oldBlock = world.getType(position).getBlock();
                 int typeId = blockstate.getTypeId();
                 int data = blockstate.getRawData();
-                int flag = ((CraftBlockState)blockstate).getFlag();
+                int flag = ((CraftBlockState) blockstate).getFlag();
                 delegate.setTypeIdAndData(x, y, z, typeId, data);
                 net.minecraft.server.Block newBlock = world.getType(position).getBlock();
                 world.notifyAndUpdatePhysics(position, null, oldBlock, newBlock, flag);
@@ -531,13 +545,15 @@ public class CraftWorld implements World {
 
     public long getTime() {
         long time = getFullTime() % 24000;
-        if (time < 0) time += 24000;
+        if (time < 0)
+            time += 24000;
         return time;
     }
 
     public void setTime(long time) {
         long margin = (time - getFullTime()) % 24000;
-        if (margin < 0) margin += 24000;
+        if (margin < 0)
+            margin += 24000;
         setFullTime(getFullTime() + margin);
     }
 
@@ -551,9 +567,11 @@ public class CraftWorld implements World {
         // Forces the client to update to the new time immediately
         for (Player p : getPlayers()) {
             CraftPlayer cp = (CraftPlayer) p;
-            if (cp.getHandle().playerConnection == null) continue;
+            if (cp.getHandle().playerConnection == null)
+                continue;
 
-            cp.getHandle().playerConnection.sendPacket(new PacketPlayOutUpdateTime(cp.getHandle().world.getTime(), cp.getHandle().getPlayerTime(), cp.getHandle().world.getGameRules().getBoolean("doDaylightCycle")));
+            cp.getHandle().playerConnection.sendPacket(new PacketPlayOutUpdateTime(cp.getHandle().world.getTime(),
+                    cp.getHandle().getPlayerTime(), cp.getHandle().world.getGameRules().getBoolean("doDaylightCycle")));
         }
     }
 
@@ -566,7 +584,7 @@ public class CraftWorld implements World {
     }
 
     public boolean createExplosion(double x, double y, double z, float power, boolean setFire, boolean breakBlocks) {
-        return !world.createExplosion(null, x, y, z, power, setFire, breakBlocks).wasCanceled;
+        return createExplosion(x, y, z, power, setFire, breakBlocks, null); // PandaSpigot 
     }
 
     public boolean createExplosion(Location loc, float power) {
@@ -577,6 +595,25 @@ public class CraftWorld implements World {
         return createExplosion(loc.getX(), loc.getY(), loc.getZ(), power, setFire);
     }
 
+    // Pandaspigot start
+    public boolean createExplosion(double x, double y, double z, float power, boolean setFire, boolean breakBlocks, Entity source) {
+        return createExplosion(x, y, z, power, setFire, breakBlocks, source, false);
+    }
+
+    public boolean createExplosion(double x, double y, double z, float power, boolean setFire, boolean breakBlocks, Entity source, boolean excludeSourceFromDamage) {
+        net.minecraft.server.Entity entity = (source == null) ? null : ((CraftEntity) source).getHandle();
+        return !world.createExplosion(entity, x, y, z, power, setFire, breakBlocks, excludeSourceFromDamage).wasCanceled;
+    }
+
+    public boolean createExplosion(Entity source, Location loc, float power, boolean setFire, boolean breakBlocks, boolean excludeSourceFromDamage) {
+        return createExplosion(loc.getX(), loc.getY(), loc.getZ(), power, setFire,breakBlocks, source, excludeSourceFromDamage);
+    }
+
+    public boolean createExplosion(Location loc, float power, boolean setFire, boolean breakBlocks, Entity source) {
+        return createExplosion(loc.getX(), loc.getY(), loc.getZ(), power, setFire, breakBlocks, source, false);
+    }
+    // PandaSpigot stop
+
     public Environment getEnvironment() {
         return environment;
     }
@@ -631,7 +668,7 @@ public class CraftWorld implements World {
 
             if (chunk != null) {
                 byte[] biomevals = chunk.getBiomeIndex();
-                biomevals[((z & 0xF) << 4) | (x & 0xF)] = (byte)bb.id;
+                biomevals[((z & 0xF) << 4) | (x & 0xF)] = (byte) bb.id;
             }
         }
     }
@@ -683,14 +720,14 @@ public class CraftWorld implements World {
     @SuppressWarnings("unchecked")
     @Deprecated
     public <T extends Entity> Collection<T> getEntitiesByClass(Class<T>... classes) {
-        return (Collection<T>)getEntitiesByClasses(classes);
+        return (Collection<T>) getEntitiesByClasses(classes);
     }
 
     @SuppressWarnings("unchecked")
     public <T extends Entity> Collection<T> getEntitiesByClass(Class<T> clazz) {
         Collection<T> list = new ArrayList<T>();
 
-        for (Object entity: world.entityList) {
+        for (Object entity : world.entityList) {
             if (entity instanceof net.minecraft.server.Entity) {
                 Entity bukkitEntity = ((net.minecraft.server.Entity) entity).getBukkitEntity();
 
@@ -712,7 +749,7 @@ public class CraftWorld implements World {
     public Collection<Entity> getEntitiesByClasses(Class<?>... classes) {
         Collection<Entity> list = new ArrayList<Entity>();
 
-        for (Object entity: world.entityList) {
+        for (Object entity : world.entityList) {
             if (entity instanceof net.minecraft.server.Entity) {
                 Entity bukkitEntity = ((net.minecraft.server.Entity) entity).getBukkitEntity();
 
@@ -740,8 +777,11 @@ public class CraftWorld implements World {
             return Collections.emptyList();
         }
 
-        AxisAlignedBB bb = new AxisAlignedBB(location.getX() - x, location.getY() - y, location.getZ() - z, location.getX() + x, location.getY() + y, location.getZ() + z);
-        List<net.minecraft.server.Entity> entityList = getHandle().a((net.minecraft.server.Entity) null, bb, null); // PAIL : rename
+        AxisAlignedBB bb = new AxisAlignedBB(location.getX() - x, location.getY() - y, location.getZ() - z,
+                location.getX() + x, location.getY() + y, location.getZ() + z);
+        List<net.minecraft.server.Entity> entityList = getHandle().a((net.minecraft.server.Entity) null, bb, null); // PAIL
+                                                                                                                    // :
+                                                                                                                    // rename
         List<Entity> bukkitEntityList = new ArrayList<org.bukkit.entity.Entity>(entityList.size());
         for (Object entity : entityList) {
             bukkitEntityList.add(((net.minecraft.server.Entity) entity).getBukkitEntity());
@@ -764,11 +804,12 @@ public class CraftWorld implements World {
     }
 
     public void save() {
-    // Spigot start
+        // Spigot start
         save(true);
     }
+
     public void save(boolean forceSave) {
-    // Spigot end
+        // Spigot end
         this.server.checkSaveState();
         try {
             boolean oldSave = world.savingDisabled;
@@ -866,18 +907,19 @@ public class CraftWorld implements World {
             Validate.isTrue(effect.getData() == null, "Wrong kind of data for this effect!");
         }
 
-        if (data != null && data.getClass().equals( org.bukkit.material.MaterialData.class )) {
+        if (data != null && data.getClass().equals(org.bukkit.material.MaterialData.class)) {
             org.bukkit.material.MaterialData materialData = (org.bukkit.material.MaterialData) data;
-            Validate.isTrue( materialData.getItemType().isBlock(), "Material must be block" );
-            spigot().playEffect( loc, effect, materialData.getItemType().getId(), materialData.getData(), 0, 0, 0, 1, 1, radius );
+            Validate.isTrue(materialData.getItemType().isBlock(), "Material must be block");
+            spigot().playEffect(loc, effect, materialData.getItemType().getId(), materialData.getData(), 0, 0, 0, 1, 1,
+                    radius);
         } else {
-            int dataValue = data == null ? 0 : CraftEffect.getDataValue( effect, data );
-            playEffect( loc, effect, dataValue, radius );
+            int dataValue = data == null ? 0 : CraftEffect.getDataValue(effect, data);
+            playEffect(loc, effect, dataValue, radius);
         }
     }
 
     public void playEffect(Location location, Effect effect, int data, int radius) {
-        spigot().playEffect( location, effect, data, 0, 0, 0, 0, 1, 1, radius );
+        spigot().playEffect(location, effect, data, 0, 0, 0, 0, 1, 1, radius);
     }
 
     public <T extends Entity> T spawn(Location location, Class<T> clazz) throws IllegalArgumentException {
@@ -886,12 +928,14 @@ public class CraftWorld implements World {
     }
 
     @Override
-    public <T extends Entity> T spawn(Location location, Class<T> clazz, java.util.function.Consumer<T> function) throws IllegalArgumentException {
+    public <T extends Entity> T spawn(Location location, Class<T> clazz, java.util.function.Consumer<T> function)
+            throws IllegalArgumentException {
         return spawn(location, clazz, function, SpawnReason.CUSTOM);
         // PandaSpigot end
     }
 
-    public FallingBlock spawnFallingBlock(Location location, org.bukkit.Material material, byte data) throws IllegalArgumentException {
+    public FallingBlock spawnFallingBlock(Location location, org.bukkit.Material material, byte data)
+            throws IllegalArgumentException {
         Validate.notNull(location, "Location cannot be null");
         Validate.notNull(material, "Material cannot be null");
         Validate.isTrue(material.isBlock(), "Material must be a block");
@@ -902,7 +946,8 @@ public class CraftWorld implements World {
 
         // PaperSpigot start - Add FallingBlock source location API
         location = location.clone();
-        EntityFallingBlock entity = new EntityFallingBlock(location, world, x, y, z, net.minecraft.server.Block.getById(material.getId()).fromLegacyData(data));
+        EntityFallingBlock entity = new EntityFallingBlock(location, world, x, y, z,
+                net.minecraft.server.Block.getById(material.getId()).fromLegacyData(data));
         // PaperSpigot end
         entity.ticksLived = 1;
 
@@ -910,12 +955,14 @@ public class CraftWorld implements World {
         return (FallingBlock) entity.getBukkitEntity();
     }
 
-    public FallingBlock spawnFallingBlock(Location location, int blockId, byte blockData) throws IllegalArgumentException {
+    public FallingBlock spawnFallingBlock(Location location, int blockId, byte blockData)
+            throws IllegalArgumentException {
         return spawnFallingBlock(location, org.bukkit.Material.getMaterial(blockId), blockData);
     }
 
     @SuppressWarnings("unchecked")
-    public net.minecraft.server.Entity createEntity(Location location, Class<? extends Entity> clazz) throws IllegalArgumentException {
+    public net.minecraft.server.Entity createEntity(Location location, Class<? extends Entity> clazz)
+            throws IllegalArgumentException {
         if (location == null || clazz == null) {
             throw new IllegalArgumentException("Location or entity class cannot be null");
         }
@@ -940,7 +987,8 @@ public class CraftWorld implements World {
             int data = blockData.getBlock().toLegacyData(blockData);
             // PaperSpigot start - Add FallingBlock source location API
             location = location.clone();
-            entity = new EntityFallingBlock(location, world, x + 0.5, y + 0.5, z + 0.5, net.minecraft.server.Block.getById(type).fromLegacyData(data));
+            entity = new EntityFallingBlock(location, world, x + 0.5, y + 0.5, z + 0.5,
+                    net.minecraft.server.Block.getById(type).fromLegacyData(data));
             // PaperSpigot end
         } else if (Projectile.class.isAssignableFrom(clazz)) {
             if (Snowball.class.isAssignableFrom(clazz)) {
@@ -957,7 +1005,8 @@ public class CraftWorld implements World {
                 entity = new EntityEnderPearl(world, null);
                 entity.setPositionRotation(x, y, z, 0, 0);
             } else if (ThrownPotion.class.isAssignableFrom(clazz)) {
-                entity = new EntityPotion(world, x, y, z, CraftItemStack.asNMSCopy(new ItemStack(org.bukkit.Material.POTION, 1)));
+                entity = new EntityPotion(world, x, y, z,
+                        CraftItemStack.asNMSCopy(new ItemStack(org.bukkit.Material.POTION, 1)));
             } else if (Fireball.class.isAssignableFrom(clazz)) {
                 if (SmallFireball.class.isAssignableFrom(clazz)) {
                     entity = new EntitySmallFireball(world);
@@ -1068,7 +1117,7 @@ public class CraftWorld implements World {
                 entity = new EntityRabbit(world);
             } else if (Endermite.class.isAssignableFrom(clazz)) {
                 entity = new EntityEndermite(world);
-            } else if (Guardian.class.isAssignableFrom(clazz)){
+            } else if (Guardian.class.isAssignableFrom(clazz)) {
                 entity = new EntityGuardian(world);
             } else if (ArmorStand.class.isAssignableFrom(clazz)) {
                 entity = new EntityArmorStand(world, x, y, z);
@@ -1092,14 +1141,16 @@ public class CraftWorld implements World {
                 height = 9;
             }
 
-            BlockFace[] faces = new BlockFace[]{BlockFace.EAST,BlockFace.NORTH,BlockFace.WEST,BlockFace.SOUTH};
+            BlockFace[] faces = new BlockFace[] { BlockFace.EAST, BlockFace.NORTH, BlockFace.WEST, BlockFace.SOUTH };
             final BlockPosition pos = new BlockPosition((int) x, (int) y, (int) z);
             for (BlockFace dir : faces) {
                 net.minecraft.server.Block nmsBlock = CraftMagicNumbers.getBlock(block.getRelative(dir));
                 if (nmsBlock.getMaterial().isBuildable() || BlockDiodeAbstract.d(nmsBlock)) {
                     boolean taken = false;
-                    AxisAlignedBB bb = EntityHanging.calculateBoundingBox(pos,CraftBlock.blockFaceToNotch(dir).opposite(),width,height);
-                    List<net.minecraft.server.Entity> list = (List<net.minecraft.server.Entity>) world.getEntities(null, bb);
+                    AxisAlignedBB bb = EntityHanging.calculateBoundingBox(pos,
+                            CraftBlock.blockFaceToNotch(dir).opposite(), width, height);
+                    List<net.minecraft.server.Entity> list = (List<net.minecraft.server.Entity>) world.getEntities(null,
+                            bb);
                     for (Iterator<net.minecraft.server.Entity> it = list.iterator(); !taken && it.hasNext();) {
                         net.minecraft.server.Entity e = it.next();
                         if (e instanceof EntityHanging) {
@@ -1115,7 +1166,7 @@ public class CraftWorld implements World {
             }
 
             EnumDirection dir = CraftBlock.blockFaceToNotch(face).opposite();
-            
+
             if (Painting.class.isAssignableFrom(clazz)) {
                 entity = new EntityPainting(world, new BlockPosition((int) x, (int) y, (int) z), dir);
             } else if (ItemFrame.class.isAssignableFrom(clazz)) {
@@ -1126,7 +1177,8 @@ public class CraftWorld implements World {
             }
 
             if (entity != null && !((EntityHanging) entity).survives()) {
-                throw new IllegalArgumentException("Cannot spawn hanging entity for " + clazz.getName() + " at " + location);
+                throw new IllegalArgumentException(
+                        "Cannot spawn hanging entity for " + clazz.getName() + " at " + location);
             }
         } else if (TNTPrimed.class.isAssignableFrom(clazz)) {
             org.bukkit.Location loc = new org.bukkit.Location(world.getWorld(), x, y, z); // PaperSpigot
@@ -1145,9 +1197,8 @@ public class CraftWorld implements World {
 
         if (entity != null) {
             // Spigot start
-            if (entity instanceof EntityOcelot)
-            {
-                ( (EntityOcelot) entity ).spawnBonus = false;
+            if (entity instanceof EntityOcelot) {
+                ((EntityOcelot) entity).spawnBonus = false;
             }
             // Spigot end
             return entity;
@@ -1157,13 +1208,15 @@ public class CraftWorld implements World {
     }
 
     @SuppressWarnings("unchecked")
-    public <T extends Entity> T addEntity(net.minecraft.server.Entity entity, SpawnReason reason) throws IllegalArgumentException {
+    public <T extends Entity> T addEntity(net.minecraft.server.Entity entity, SpawnReason reason)
+            throws IllegalArgumentException {
         // PandaSpigot start - function param
         return addEntity(entity, reason, null);
     }
 
     @SuppressWarnings("unchecked")
-    public <T extends Entity> T addEntity(net.minecraft.server.Entity entity, SpawnReason reason, java.util.function.Consumer<T> function) throws IllegalArgumentException {
+    public <T extends Entity> T addEntity(net.minecraft.server.Entity entity, SpawnReason reason,
+            java.util.function.Consumer<T> function) throws IllegalArgumentException {
         // PandaSpigot end
         Preconditions.checkArgument(entity != null, "Cannot spawn null entity");
 
@@ -1180,12 +1233,14 @@ public class CraftWorld implements World {
         return (T) entity.getBukkitEntity();
     }
 
-    public <T extends Entity> T spawn(Location location, Class<T> clazz, SpawnReason reason) throws IllegalArgumentException {
+    public <T extends Entity> T spawn(Location location, Class<T> clazz, SpawnReason reason)
+            throws IllegalArgumentException {
         // PandaSpigot start - function param
         return spawn(location, clazz, null, reason);
     }
 
-    public <T extends Entity> T spawn(Location location, Class<T> clazz, java.util.function.Consumer<T> function, SpawnReason reason) throws IllegalArgumentException {
+    public <T extends Entity> T spawn(Location location, Class<T> clazz, java.util.function.Consumer<T> function,
+            SpawnReason reason) throws IllegalArgumentException {
         // PandaSpigot end
         net.minecraft.server.Entity entity = createEntity(location, clazz);
 
@@ -1373,9 +1428,9 @@ public class CraftWorld implements World {
         ambientSpawn = limit;
     }
 
-
     public void playSound(Location loc, Sound sound, float volume, float pitch) {
-        if (loc == null || sound == null) return;
+        if (loc == null || sound == null)
+            return;
 
         double x = loc.getX();
         double y = loc.getY();
@@ -1390,9 +1445,11 @@ public class CraftWorld implements World {
 
     public boolean setGameRuleValue(String rule, String value) {
         // No null values allowed
-        if (rule == null || value == null) return false;
+        if (rule == null || value == null)
+            return false;
 
-        if (!isGameRule(rule)) return false;
+        if (!isGameRule(rule))
+            return false;
 
         getHandle().getGameRules().set(rule, value);
         return true;
@@ -1442,93 +1499,80 @@ public class CraftWorld implements World {
             cps.queueUnload(chunk.locX, chunk.locZ);
         }
     }
+
     // Spigot start
-    private final Spigot spigot = new Spigot()
-    {
+    private final Spigot spigot = new Spigot() {
         @Override
-        public void playEffect( Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius )
-        {
-            Validate.notNull( location, "Location cannot be null" );
-            Validate.notNull( effect, "Effect cannot be null" );
-            Validate.notNull( location.getWorld(), "World cannot be null" );
+        public void playEffect(Location location, Effect effect, int id, int data, float offsetX, float offsetY,
+                float offsetZ, float speed, int particleCount, int radius) {
+            Validate.notNull(location, "Location cannot be null");
+            Validate.notNull(effect, "Effect cannot be null");
+            Validate.notNull(location.getWorld(), "World cannot be null");
             Packet packet;
-            if ( effect.getType() != Effect.Type.PARTICLE )
-            {
+            if (effect.getType() != Effect.Type.PARTICLE) {
                 int packetData = effect.getId();
-                packet = new PacketPlayOutWorldEvent( packetData, new BlockPosition(location.getBlockX(), location.getBlockY(), location.getBlockZ() ), id, false );
-            } else
-            {
+                packet = new PacketPlayOutWorldEvent(packetData,
+                        new BlockPosition(location.getBlockX(), location.getBlockY(), location.getBlockZ()), id, false);
+            } else {
                 net.minecraft.server.EnumParticle particle = null;
                 int[] extra = null;
-                for ( net.minecraft.server.EnumParticle p : net.minecraft.server.EnumParticle.values() )
-                {
-                    if ( effect.getName().startsWith( p.b().replace("_", "") ) )
-                    {
+                for (net.minecraft.server.EnumParticle p : net.minecraft.server.EnumParticle.values()) {
+                    if (effect.getName().startsWith(p.b().replace("_", ""))) {
                         particle = p;
-                        if ( effect.getData() != null ) 
-                        {
-                            if ( effect.getData().equals( org.bukkit.Material.class ) )
-                            {
-                                extra = new int[]{ id };
-                            } else 
-                            {
-                                extra = new int[]{ (data << 12) | (id & 0xFFF) };
+                        if (effect.getData() != null) {
+                            if (effect.getData().equals(org.bukkit.Material.class)) {
+                                extra = new int[] { id };
+                            } else {
+                                extra = new int[] { (data << 12) | (id & 0xFFF) };
                             }
                         }
                         break;
                     }
                 }
-                if ( extra == null )
-                {
+                if (extra == null) {
                     extra = new int[0];
                 }
-                packet = new PacketPlayOutWorldParticles( particle, true, (float) location.getX(), (float) location.getY(), (float) location.getZ(), offsetX, offsetY, offsetZ, speed, particleCount, extra );
+                packet = new PacketPlayOutWorldParticles(particle, true, (float) location.getX(),
+                        (float) location.getY(), (float) location.getZ(), offsetX, offsetY, offsetZ, speed,
+                        particleCount, extra);
             }
             int distance;
             radius *= radius;
-            for ( Player player : getPlayers() )
-            {
-                if ( ( (CraftPlayer) player ).getHandle().playerConnection == null )
-                {
+            for (Player player : getPlayers()) {
+                if (((CraftPlayer) player).getHandle().playerConnection == null) {
                     continue;
                 }
-                if ( !location.getWorld().equals( player.getWorld() ) )
-                {
+                if (!location.getWorld().equals(player.getWorld())) {
                     continue;
                 }
-                distance = (int) player.getLocation().distanceSquared( location );
-                if ( distance <= radius )
-                {
-                    ( (CraftPlayer) player ).getHandle().playerConnection.sendPacket( packet );
+                distance = (int) player.getLocation().distanceSquared(location);
+                if (distance <= radius) {
+                    ((CraftPlayer) player).getHandle().playerConnection.sendPacket(packet);
                 }
             }
         }
 
         @Override
-        public void playEffect( Location location, Effect effect )
-        {
-            CraftWorld.this.playEffect( location, effect, 0 );
+        public void playEffect(Location location, Effect effect) {
+            CraftWorld.this.playEffect(location, effect, 0);
         }
 
         @Override
-        public LightningStrike strikeLightning(Location loc, boolean isSilent)
-        {
-            EntityLightning lightning = new EntityLightning( world, loc.getX(), loc.getY(), loc.getZ(), false, isSilent );
-            world.strikeLightning( lightning );
-            return new CraftLightningStrike( server, lightning );
+        public LightningStrike strikeLightning(Location loc, boolean isSilent) {
+            EntityLightning lightning = new EntityLightning(world, loc.getX(), loc.getY(), loc.getZ(), false, isSilent);
+            world.strikeLightning(lightning);
+            return new CraftLightningStrike(server, lightning);
         }
 
         @Override
-        public LightningStrike strikeLightningEffect(Location loc, boolean isSilent)
-        {
-            EntityLightning lightning = new EntityLightning( world, loc.getX(), loc.getY(), loc.getZ(), true, isSilent );
-            world.strikeLightning( lightning );
-            return new CraftLightningStrike( server, lightning );
+        public LightningStrike strikeLightningEffect(Location loc, boolean isSilent) {
+            EntityLightning lightning = new EntityLightning(world, loc.getX(), loc.getY(), loc.getZ(), true, isSilent);
+            world.strikeLightning(lightning);
+            return new CraftLightningStrike(server, lightning);
         }
     };
 
-    public Spigot spigot()
-    {
+    public Spigot spigot() {
         return spigot;
     }
     // Spigot end
