From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MasterDash5 <github@masterdash5.com>
Date: Thu, 30 Jul 2026 16:37:49 -0600
Subject: [PATCH] Fix bottles moving slots when using a cauldron


diff --git a/src/main/java/net/minecraft/server/BlockCauldron.java b/src/main/java/net/minecraft/server/BlockCauldron.java
index 30344761ff4d078d1c5a2066c773f96534368c49..651fbf18782bae8b4474e471fefdab3c1a8aa5b6 100644
--- a/src/main/java/net/minecraft/server/BlockCauldron.java
+++ b/src/main/java/net/minecraft/server/BlockCauldron.java
@@ -85,17 +85,22 @@ public class BlockCauldron extends Block {
                         if (!entityhuman.abilities.canInstantlyBuild) {
                             ItemStack itemstack1 = new ItemStack(Items.POTION, 1, 0);
 
-                            if (!entityhuman.inventory.pickup(itemstack1)) {
+                            // PandaSpigot start - Fix bottles moving slots when using a cauldron
+                            if (itemstack.count <= 1) {
+                                entityhuman.inventory.setItem(entityhuman.inventory.itemInHandIndex, itemstack1);
+                            } else if (!entityhuman.inventory.pickup(itemstack1)) {
+                            // PandaSpigot end
                                 world.addEntity(new EntityItem(world, (double) blockposition.getX() + 0.5D, (double) blockposition.getY() + 1.5D, (double) blockposition.getZ() + 0.5D, itemstack1));
                             } else if (entityhuman instanceof EntityPlayer) {
                                 ((EntityPlayer) entityhuman).updateInventory(entityhuman.defaultContainer);
                             }
 
                             entityhuman.b(StatisticList.J);
-                            --itemstack.count;
-                            if (itemstack.count <= 0) {
-                                entityhuman.inventory.setItem(entityhuman.inventory.itemInHandIndex, (ItemStack) null);
+                            // PandaSpigot start - Fix bottles moving slots when using a cauldron
+                            if (itemstack.count > 1) {
+                                --itemstack.count;
                             }
+                            // PandaSpigot end
                         }
 
                         this.a(world, blockposition, iblockdata, i - 1);
